src/Chainweb/Cut.hs

```diff
@@ -185,42 +201,115 @@ cutWeight :: Getter Cut BlockWeight
+    , _cutChainwebVersion = v
+    }
 
-cutAdjPairs :: Getter Cut (HS.HashSet (AdjPair BlockHeader))
-cutAdjPairs = to _cutAdjPairs
-{-# INLINE cutAdjPairs #-}
+-- -------------------------------------------------------------------------- --
+-- Adjacents
+--
+-- Note that in a cut adjacency is directed. The reason is that the headers in a
+-- cut may be of different height and may thus use different graphs.
 
--- | Assumes that the cid exists in the graph.
+-- | Map of adjacent block headers for a chain in a cut. The function considers
+-- the block height for the respective chain in the given cut.
+--
+-- Assumes that the cid exists in the graph.
 --
 cutAdjs
     :: HasChainId cid
     => Cut
     -> cid
     -> HM.HashMap ChainId BlockHeader
-cutAdjs c cid = HM.intersection
-    (_cutHeaders c)
-    (HS.toMap (adjacentChainIds (_chainGraph c) cid))
+cutAdjs c = HM.intersection (_cutHeaders c) . HS.toMap . cutAdjChainIds c
+{-# INLINE cutAdjs #-}
 
-unsafeMkCut :: ChainwebVersion -> HM.HashMap ChainId BlockHeader -> Cut
-unsafeMkCut v hdrs = Cut
-    { _cutHeaders = hdrs
-    , _cutChainwebVersion = v
-    }
+-- | Adjacent chain ids for a chain in a cut. The function considers the block
+-- height for the respective chain in the given cut.
+--
+-- Assumes that the cid exists in the graph.
+--
+cutAdjChainIds :: HasChainId cid => Cut -> cid -> HS.HashSet ChainId
+cutAdjChainIds c cid = c ^?! ixg (_chainId cid) . blockAdjacentChainIds
+{-# INLINE cutAdjChainIds #-}
+
+-- | /Directed/ adjacent pairs of a cut. Note that block headers in a cut can be
+-- of different block height and can thus use different chain graphs.
+--
+_cutAdjPairs :: Cut -> [(BlockHeader, BlockHeader)]
+_cutAdjPairs c =
+    [ (h, x)
+    | (cid, h) <- HM.toList (_cutMap c)
+    , x <- toList (cutAdjs c cid)
+    ]
+{-# INLINE _cutAdjPairs #-}
+
+cutAdjPairs :: Getter Cut [(BlockHeader, BlockHeader)]
+cutAdjPairs = to _cutAdjPairs
+{-# INLINE cutAdjPairs #-}
+
+-- -------------------------------------------------------------------------- --
+-- Chain Heights
+
+chainHeights :: Cut -> [BlockHeight]
+chainHeights = fmap (_blockHeight) . toList . _cutHeaders
+{-# INLINE chainHeights #-}
+
+meanChainHeight :: Cut -> BlockHeight
+meanChainHeight = mean . chainHeights
+  where
+    mean l = round $ sum @_ @Double (realToFrac <$> l) / realToFrac (length l)
+{-# INLINE meanChainHeight #-}
+
+maxChainHeight :: Cut -> BlockHeight
+maxChainHeight = maximum . chainHeights
+{-# INLINE maxChainHeight #-}
+
+minChainHeight :: Cut -> BlockHeight
+minChainHeight = minimum . chainHeights
+{-# INLINE minChainHeight #-}
+
+-- | Returns whether a chain graph transition occurs within the cut.
+--
+isTransitionCut :: Cut -> Bool
+isTransitionCut c = minChainHeight c < lastGraphChange c (maxChainHeight c)
 
 -- -------------------------------------------------------------------------- --
 -- Limit Cut Hashes By Height
 
+-- | This an internal function. The result is meaningful only if the input
```

Comment:
Should it be `projectChainsUnsafe`?

Reply:
I think that might be overkill. This module is full of invariants about the internal structure of cuts. Most likely, we would have to prefix all non-exported functions with "unsafe" :-)